### ISSUE TITLE
New lint: unused-async

### ISSUE THREAD

**Issue by InquisitivePenguin** [rust-lang/rust-clippy#7176]
### What it does

Checks for functions that are declared `async` but have no `.await`s inside of them.

### Categories (optional)

- Kind: pedantic

Reduces code smell and overhead by encouraging async functions to be refactored to synchronous functions.

### Drawbacks
There are valid cases where a programmer may want to wrap synchronous code in an asynchronous future: for example, spawning a synchronous task on an asynchronous scheduler, or implementing an async function for a trait where the implementation is synchronous. These cases are the reason this lint is pedantic. However, there are a few additional steps I want to take with this lint to avoid causing hassle or confusion for programmers:
- This lint should avoid trait functions, as there are valid reasons why an async trait function might have a non-async implementation.
- This lint should recommend to the programmer the proper way to make an synchronous task into a future: using the `async {}` block.

### Example

```rust
async fn get_random_number() -> i64 {
    4 // Chosen by fair dice roll. Guaranteed to be random.
}
```

Could be written as:

```rust
fn get_random_number() -> i64 {
    4 // Chosen by fair dice roll. Guaranteed to be random.
}
```

Something like this, however, should **not** be caught by clippy:
```rust
#[async_trait]
trait AsyncTrait {
    async fn foo();
}

struct Bar;

#[async_trait]
impl AsyncTrait for Bar {
    async fn foo() {
        println!("bar");
    }
}
```

**Comment by InquisitivePenguin** [comment]
I'd like to implement this myself, if possible. Please let me know if this lint is a bad idea, or if you have any suggestions for improvement.

@rustbot claim
